<a href="https://colab.research.google.com/github/henryezikeai/Prediction-of-product-sales/blob/main/1692960172037_Project_1__Part_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Import packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
pd.set_option('display.max_columns',100)
import missingno
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.pipeline import make_pipeline
from sklearn.impute import SimpleImputer
# Set pandas as the default output for sklearn
from sklearn import set_config
set_config(transform_output='pandas')
from sklearn.model_selection import train_test_split, GridSearchCV


In [ ]:
filepath = "/content/drive/MyDrive/CodingDojo/01-Fundamentals/Week02/Data/sales_predictions_2023.csv"

In [ ]:
df = pd.read_csv(filepath)

In [ ]:
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [ ]:
df.isna().sum() #Checking for missing items

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [ ]:
replacements = {
    'LF':'Low Fat',
    'low fat':'Low Fat',
    'reg':'Regular'
}

#Created a dictionary for the replacement of inconsistent values

In [ ]:
val = df['Item_Fat_Content'].value_counts()
val

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

In [ ]:
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace(replacements) #Replaced inconsistent values with values that are consistent


In [ ]:
df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

In [ ]:
df = df.drop(columns='Item_Identifier')

In [ ]:
y = df['Item_Outlet_Sales'] #Assigned my y value


In [ ]:
X = df.drop(columns='Item_Outlet_Sales') #Created the features

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42) #Did a split between training and test data.

In [ ]:
num_cols = X_train.select_dtypes("number").columns


In [ ]:
impute_median = SimpleImputer(strategy='median') #Instantiated SimpleImputer

In [ ]:
scaler = StandardScaler() #Instantiated StandardScaler

In [ ]:
num_pipe  = make_pipeline(impute_median, scaler) #Created a pipeline for numeric values

In [ ]:
num_pipe

Pipeline(steps=[('simpleimputer', SimpleImputer(strategy='median')),
                ('standardscaler', StandardScaler())])

In [ ]:
cat_cols = X_train.select_dtypes("object").columns

In [ ]:
impute_constant = SimpleImputer(strategy='constant', fill_value="Missing")

In [ ]:
ohe_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')

In [ ]:
cat_pipe = make_pipeline(impute_constant, ohe_encoder)

In [ ]:
cat_pipe

Pipeline(steps=[('simpleimputer',
                 SimpleImputer(fill_value='Missing', strategy='constant')),
                ('onehotencoder',
                 OneHotEncoder(handle_unknown='ignore', sparse_output=False))])

In [ ]:
num_tuple = ('numeric', num_pipe, num_cols)

In [ ]:
cat_tuple = ('categorical', cat_pipe, cat_cols)

In [ ]:
col_transformer = ColumnTransformer([num_tuple, cat_tuple], verbose_feature_names_out = False)

In [ ]:
col_transformer

ColumnTransformer(transformers=[('numeric',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='median')),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                ('categorical',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(fill_value='Missing',
                                                                strategy='constant')),
                                                 ('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier', 'Outlet_Size',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object'))],
                  verbose_feature_names_out=False)

In [ ]:
col_transformer.fit(X_train)

ColumnTransformer(transformers=[('numeric',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(strategy='median')),
                                                 ('standardscaler',
                                                  StandardScaler())]),
                                 Index(['Item_Weight', 'Item_Visibility', 'Item_MRP',
       'Outlet_Establishment_Year'],
      dtype='object')),
                                ('categorical',
                                 Pipeline(steps=[('simpleimputer',
                                                  SimpleImputer(fill_value='Missing',
                                                                strategy='constant')),
                                                 ('onehotencoder',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False))]),
                                 Index(['Item_Fat_Content', 'Item_Type', 'Outlet_Identifier', 'Outlet_Size',
       'Outlet_Location_Type', 'Outlet_Type'],
      dtype='object'))],
                  verbose_feature_names_out=False)

In [ ]:
X_train_processed = col_transformer.transform(X_train) #Transformed Training data

In [ ]:
X_test_processed = col_transformer.transform(X_test) #Transformed Testing data

In [ ]:
X_train_processed

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
4776     0.827485        -0.712775  1.828109                   1.327849   
7510     0.566644        -1.291052  0.603369                   1.327849   
5828    -0.121028         1.813319  0.244541                   0.136187   
5327    -1.158464        -1.004931 -0.952591                   0.732018   
4810     1.538870        -0.965484 -0.336460                   0.493686   
...           ...              ...       ...                        ...   
5734    -0.821742         4.309657 -0.044657                   0.017021   
5191     0.649639         1.008625 -1.058907                   1.089517   
5390     1.123896        -0.920527  1.523027                   0.493686   
860      1.775999        -0.227755 -0.383777                   1.089517   
7270     0.827485        -0.958677 -0.738361                  -0.102145   

      Item_Fat_Content_Low Fat  Item_Fat_Content_Regular  \
4776                       1.0                       0.0   
7510                       0.0                       1.0   
5828                       0.0                       1.0   
5327                       1.0                       0.0   
4810                       1.0                       0.0   
...                        ...                       ...   
5734                       0.0                       1.0   
5191                       1.0                       0.0   
5390                       1.0                       0.0   
860                        1.0                       0.0   
7270                       1.0                       0.0   

      Item_Type_Baking Goods  Item_Type_Breads  Item_Type_Breakfast  \
4776                     0.0               0.0                  0.0   
7510                     0.0               0.0                  0.0   
5828                     0.0               0.0                  0.0   
5327                     1.0               0.0                  0.0   
4810                     0.0               0.0                  0.0   
...                      ...               ...                  ...   
5734                     0.0               0.0                  0.0   
5191                     0.0               0.0                  0.0   
5390                     0.0               0.0                  0.0   
860                      0.0               0.0                  0.0   
7270                     0.0               0.0                  0.0   

      Item_Type_Canned  Item_Type_Dairy  Item_Type_Frozen Foods  \
4776               0.0              0.0                     0.0   
7510               0.0              0.0                     0.0   
5828               0.0              0.0                     0.0   
5327               0.0              0.0                     0.0   
4810               0.0              0.0                     1.0   
...                ...              ...                     ...   
5734               0.0              0.0                     0.0   
5191               0.0              0.0                     1.0   
5390               0.0              0.0                     0.0   
860                0.0              0.0                     0.0   
7270               0.0              0.0                     0.0   

      Item_Type_Fruits and Vegetables  Item_Type_Hard Drinks  \
4776                              0.0                    0.0   
7510                              0.0                    0.0   
5828                              0.0                    0.0   
5327                              0.0                    0.0   
4810                              0.0                    0.0   
...                               ...                    ...   
5734                              1.0                    0.0   
5191                              0.0                    0.0   
5390                              0.0                    0.0   
860                               0.0                    0.0   
7270                  

In [ ]:
def regression_metrics(y_true, y_pred, label='', verbose = True, output_dict=False):
  # Get metrics
  mae = mean_absolute_error(y_true, y_pred)
  mse = mean_squared_error(y_true, y_pred)
  rmse = mean_squared_error(y_true, y_pred, squared=False)
  r_squared = r2_score(y_true, y_pred)
  if verbose == True:
    # Print Result with Label and Header
    header = "-"*60
    print(header, f"Regression Metrics: {label}", header, sep='\n')
    print(f"- MAE = {mae:,.3f}")
    print(f"- MSE = {mse:,.3f}")
    print(f"- RMSE = {rmse:,.3f}")
    print(f"- R^2 = {r_squared:,.3f}")
  if output_dict == True:
      metrics = {'Label':label, 'MAE':mae,
                 'MSE':mse, 'RMSE':rmse, 'R^2':r_squared}
      return metrics

def evaluate_regression(reg, X_train, y_train, X_test, y_test, verbose = True,
                        output_frame=False):
  # Get predictions for training data
  y_train_pred = reg.predict(X_train)

  # Call the helper function to obtain regression metrics for training data
  results_train = regression_metrics(y_train, y_train_pred, verbose = verbose,
                                     output_dict=output_frame,
                                     label='Training Data')
  print()
  # Get predictions for test data
  y_test_pred = reg.predict(X_test)
  # Call the helper function to obtain regression metrics for test data
  results_test = regression_metrics(y_test, y_test_pred, verbose = verbose,
                                  output_dict=output_frame,
                                    label='Test Data' )

  # Store results in a dataframe if ouput_frame is True
  if output_frame:
    results_df = pd.DataFrame([results_train,results_test])
    # Set the label as the index
    results_df = results_df.set_index('Label')
    # Set index.name to none to get a cleaner looking result
    results_df.index.name=None
    # Return the dataframe
    return results_df.round(3)



In [ ]:
linreg = LinearRegression()
linreg.fit(X_train_processed, y_train)
evaluate_regression(linreg, X_train_processed, y_train, X_test_processed, y_test )

------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 847.160
- MSE = 1,297,557.253
- RMSE = 1,139.104
- R^2 = 0.562

------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 804.104
- MSE = 1,194,262.641
- RMSE = 1,092.823
- R^2 = 0.567


**I'd say that the model is neither overfit or underfit. It needs more data. The reason I conclude that it is neither overfit or underfit is because 0.567 and 0.562 are not poor. I'd call it average perfomance.**

In [ ]:
rf = RandomForestRegressor(random_state=42)
rf.fit(X_train_processed, y_train)
evaluate_regression(rf, X_train_processed, y_train, X_test_processed, y_test )

------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 296.340
- MSE = 182,514.168
- RMSE = 427.217
- R^2 = 0.938

------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 766.805
- MSE = 1,218,314.500
- RMSE = 1,103.773
- R^2 = 0.558


**The model is overfit. It's performing way better on training data than on test data. Linear Regression model also performed better on Test data.**

In [ ]:
rf.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'criterion': 'squared_error',
 'max_depth': None,
 'max_features': 1.0,
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': 42,
 'verbose': 0,
 'warm_start': False}

In [ ]:
param_grid = {'max_depth': [None],
              'n_estimators':[500],
              'min_samples_leaf':[9],
              'min_samples_split':[2,10]
              }

In [ ]:
gridsearch = GridSearchCV(rf, param_grid, n_jobs=-1, cv = 3, verbose=1)
gridsearch.fit(X_train_processed, y_train)

Fitting 3 folds for each of 2 candidates, totalling 6 fits


GridSearchCV(cv=3, estimator=RandomForestRegressor(random_state=42), n_jobs=-1,
             param_grid={'max_depth': [None], 'min_samples_leaf': [9],
                         'min_samples_split': [2, 10], 'n_estimators': [500]},
             verbose=1)

In [ ]:
best_rf = gridsearch.best_estimator_
evaluate_regression(best_rf, X_train_processed, y_train, X_test_processed, y_test)

------------------------------------------------------------
Regression Metrics: Training Data
------------------------------------------------------------
- MAE = 633.414
- MSE = 831,962.048
- RMSE = 912.120
- R^2 = 0.719

------------------------------------------------------------
Regression Metrics: Test Data
------------------------------------------------------------
- MAE = 736.893
- MSE = 1,125,934.910
- RMSE = 1,061.101
- R^2 = 0.592


##4. Evaluation

**- Overall, I'll recommend the Random Forst Model. It did better than Linear Regression on test data on every metric.**

Question 1: Interpret your model's performance based on R-squared in a way that your non-technical stakeholder can understand.

- The R2 Squared metric of 0.592 means that our model is capable of predicting the correct value 59.2% of the time. The performance of the model might be average but it is better than the Linear Regression model. We always want R2 to be as close to 1/100% as possible on both the training data and test data but especially the test data.

Question 2: Select another regression metric (RMSE/MAE/MSE) to express the performance of your model to your stakeholder.

- The MAE is another metric that we can compare for both models. The Linear Regression model scored a MAE of 804.10 on test data and Random Forest model scored 736.89. If we are to make a decision on what model to deploy based on these metrics, I would encourage the business unit to deploy the Random Forest Model. Because it has a lower MAE. Generally, we want our models to make less errors. When our models make less errors, it gets maore predictions correctly which translates into more profits for the business.

Question 3: Include why you selected this metric to explain to your stakeholder.

- I selected the MAE to explain to the stakeholders because it is simpler to understand than RMSE and MSE.

Question 4: Compare the training vs. test scores and answer the question: to what extent is this model overfit/underfit?

- By comparing the training data metrics and the test data metrics of the Random Forest Model, I would say that the model is neither overfit or underfit. It's not underfit because it has performed average on both training data and test data.
It's also not overfit because the training data metrics is doing doing way better than the test data metrics. The differences are minor.  

